File: BubblerApp/BubblerApp/Features/Graph/GraphFeedViewModel.swift
import Combine
import Foundation

@MainActor
final class GraphFeedViewModel: ObservableObject {
    @Published private(set) var currentNode: GraphFeedNode?
    @Published private(set) var nextChoices: [GraphFeedNode] = []
    @Published private(set) var isLoading = false
    @Published private(set) var isSubmitting = false
    @Published private(set) var isCurrentPostLiked = false
    @Published private(set) var seedStrategyLabel: String?
    @Published var errorMessage: String?
    @Published var statusMessage: String?

    private var preferences = UserPreferences.placeholder
    private var sessionQueue: [GraphFeedNode] = []
    private var currentPostStartedAt: Date?
    private var likedPostIDs = Set<String>()

    var hasCurrentPost: Bool {
        currentNode != nil
    }

    var currentTopicName: String {
        currentNode?.topicName ?? "Topicless bubble"
    }

    var hasCurrentTopic: Bool {
        currentNode?.topicName != nil
    }

    var isCurrentTopicPreferred: Bool {
        currentNode?.isPreferredTopic == true
    }

    var isCurrentTopicBlacklisted: Bool {
        currentNode?.isBlacklistedTopic == true
    }

    func load(using authSession: AuthSession) async {
        guard !isLoading else { return }
        await loadSession(
            using: authSession,
            diversify: false,
            message: "Building your graph session."
        )
    }

    func refreshSession(using authSession: AuthSession) async {
        await loadSession(
            using: authSession,
            diversify: true,
            message: "Exploring a fresh path across topics."
        )
    }

    func choose(_ node: GraphFeedNode, using authSession: AuthSession) async {
        await advance(
            byRecording: .explore,
            explicitNextNode: node,
            using: authSession,
            fallbackMessage: "Following a connected post."
        )
    }

    func likeCurrentPost(using authSession: AuthSession) async {
        guard let currentNode else {
            await loadSession(
                using: authSession,
                diversify: false,
                message: "Building your graph session."
            )
            return
        }

        isSubmitting = true
        errorMessage = nil

        do {
            try await recordInteraction(for: currentNode, type: .like)
            likedPostIDs.insert(currentNode.id)
            isCurrentPostLiked = true
            statusMessage = "Liked this post."
        } catch {
            handle(error, using: authSession, fallbackMessage: "We couldn't save that like.")
        }

        isSubmitting = false
    }

    func unlikeCurrentPost(using authSession: AuthSession) async {
        guard let currentNode else { return }

        isSubmitting = true
        errorMessage = nil

        do {
            try await APIClient.deleteLike(postId: currentNode.id)
            likedPostIDs.remove(currentNode.id)
            isCurrentPostLiked = false
            statusMessage = "Removed your like."
        } catch {
            handle(error, using: authSession, fallbackMessage: "We couldn't remove that like.")
        }

        isSubmitting = false
    }

    func toggleCurrentPostLike(using authSession: AuthSession) async {
        if isCurrentPostLiked {
            await unlikeCurrentPost(using: authSession)
        } else {
            await likeCurrentPost(using: authSession)
        }
    }

    func skipCurrentPost(using authSession: AuthSession) async {
        await advance(
            byRecording: .skip,
            explicitNextNode: nil,
            using: authSession,
            fallbackMessage: "Skipping ahead to the next bubble."
        )
    }

    func togglePreferCurrentTopic(using authSession: AuthSession) async {
        guard let topic = currentNode?.topicName else { return }

        isSubmitting = true
        errorMessage = nil

        do {
            var updated = preferences
            if contains(normalizedTopicName(from: topic), in: updated.preferredTopics) {
                updated.unpreferTopic(topic)
                statusMessage = "Removed \(topic) from preferred topics."
            } else {
                updated.preferTopic(topic)
                statusMessage = "Preferred topic: \(topic)."
            }

            preferences = try await APIClient.updatePreferences(updated.sanitized().updatePayload).sanitized()
            refreshPreferenceFlags()
        } catch {
            handle(error, using: authSession, fallbackMessage: "We couldn't update topic preferences.")
        }

        isSubmitting = false
    }

    func toggleBlacklistCurrentTopic(using authSession: AuthSession) async {
        guard let currentNode,
              let topic = currentNode.topicName else {
            return
        }

        isSubmitting = true
        errorMessage = nil

        do {
            var updated = preferences
            let wasBlacklisted = contains(
                normalizedTopicName(from: topic),
                in: updated.blacklistedTopics
            )

            if wasBlacklisted {
                updated.unblacklistTopic(topic)
                preferences = try await APIClient.updatePreferences(updated.sanitized().updatePayload).sanitized()
                refreshPreferenceFlags()
                statusMessage = "Removed \(topic) from blacklist."
            } else {
                updated.blacklistTopic(topic)
                preferences = try await APIClient.updatePreferences(updated.sanitized().updatePayload).sanitized()
                nextChoices = []
                sessionQueue = []

                try await recordInteraction(for: currentNode, type: .skip)

                await loadSession(
                    using: authSession,
                    diversify: true,
                    message: "Blacklisted \(topic). Exploring other bubbles."
                )
            }
        } catch {
            handle(error, using: authSession, fallbackMessage: "We couldn't update topic preferences.")
        }

        isSubmitting = false
    }

    func updateCurrentPostContent(_ content: String) {
        guard let currentNode else { return }
        let existing = currentNode.post
        let updatedPost = Post(
            id: existing.id,
            userId: existing.userId,
            content: content,
            createdAt: existing.createdAt,
            topic: existing.topic,
            embedding: existing.embedding
        )
        self.currentNode = GraphFeedNode(
            post: updatedPost,
            isPreferredTopic: currentNode.isPreferredTopic,
            isBlacklistedTopic: currentNode.isBlacklistedTopic
        )
    }

    func deleteCurrentPost(using authSession: AuthSession) async {
        guard let currentNode else { return }

        isSubmitting = true
        errorMessage = nil

        do {
            try await APIClient.deletePost(id: currentNode.id)
            authSession.showSuccessMessage("Post deleted.")

            likedPostIDs.remove(currentNode.id)
            nextChoices.removeAll { $0.id == currentNode.id }
            sessionQueue.removeAll { $0.id == currentNode.id }

            if let nextNode = nextAutomaticNode(excluding: currentNode.id) {
                await setCurrentNode(nextNode, using: authSession)
                if errorMessage == nil {
                    statusMessage = "Deleted your post and moved ahead."
                }
            } else {
                await loadSession(
                    using: authSession,
                    diversify: true,
                    message: "Deleted your post. Exploring other bubbles."
                )
            }
        } catch {
            handle(error, using: authSession, fallbackMessage: "We couldn't delete that post.")
        }

        isSubmitting = false
    }

    func viewTimeText(at date: Date) -> String {
        let elapsedSeconds = Int(viewTime(at: date).rounded(.down))
        return "\(elapsedSeconds)s tracked"
    }

    private func loadSession(
        using authSession: AuthSession,
        diversify: Bool = false,
        message: String
    ) async {
        isLoading = true
        errorMessage = nil
        statusMessage = message

        do {
            preferences = try await APIClient.getPreferences().sanitized()
            await refreshLikedPostIDs()

            let sessionNodes = try await fetchUsableSessionNodes(diversify: diversify)
            guard let firstNode = sessionNodes.first else {
                throw GraphFeedError.noUsablePosts
            }

            sessionQueue = Array(sessionNodes.dropFirst())
            await setCurrentNode(firstNode, using: authSession)

            if errorMessage == nil {
                let seedNote = seedStrategyLabel.map { " \($0)." } ?? ""
                statusMessage = statusMessage(
                    for: firstNode,
                    defaultMessage: "Session ready.\(seedNote)"
                )
            }
        } catch {
            currentNode = nil
            nextChoices = []
            sessionQueue = []
            isCurrentPostLiked = false
            seedStrategyLabel = nil
            handle(error, using: authSession, fallbackMessage: "We couldn't build a graph session right now.")
        }

        isLoading = false
    }

    private func advance(
        byRecording interactionType: GraphInteractionType,
        explicitNextNode: GraphFeedNode?,
        using authSession: AuthSession,
        fallbackMessage: String
    ) async {
        guard let currentNode else {
            await loadSession(
                using: authSession,
                diversify: false,
                message: "Building your graph session."
            )
            return
        }

        isSubmitting = true
        errorMessage = nil

        do {
            try await recordInteraction(for: currentNode, type: interactionType)

            if let explicitNextNode {
                await setCurrentNode(explicitNextNode, using: authSession)
                if errorMessage == nil {
                    statusMessage = statusMessage(for: explicitNextNode, defaultMessage: fallbackMessage)
                }
            } else if let nextNode = nextAutomaticNode(excluding: currentNode.id) {
                await setCurrentNode(nextNode, using: authSession)
                if errorMessage == nil {
                    statusMessage = statusMessage(for: nextNode, defaultMessage: fallbackMessage)
                }
            } else {
                await loadSession(
                    using: authSession,
                    diversify: true,
                    message: "Loading a fresh cross-topic path."
                )
            }
        } catch {
            handle(error, using: authSession, fallbackMessage: "We couldn't save that interaction.")
        }

        isSubmitting = false
    }

    private func setCurrentNode(_ node: GraphFeedNode, using authSession: AuthSession) async {
        guard !node.isBlacklistedTopic else {
            await loadSession(
                using: authSession,
                diversify: true,
                message: "That topic is blacklisted, so exploring other bubbles."
            )
            return
        }

        currentNode = node
        currentPostStartedAt = Date()
        isCurrentPostLiked = likedPostIDs.contains(node.id)

        do {
            nextChoices = try await loadChoices(for: node)

            if nextChoices.isEmpty, sessionQueue.isEmpty {
                statusMessage = "No connected posts were available, so the next action will pull a fresh session."
            }
        } catch {
            nextChoices = []
            handle(error, using: authSession, fallbackMessage: "We couldn't load the connected posts.")
        }
    }

    private func fetchUsableSessionNodes(
        diversify: Bool,
        maxAttempts: Int = 3
    ) async throws -> [GraphFeedNode] {
        var lastSeedLabel: String?

        for attempt in 0 ..< maxAttempts {
            // Escalate to diversify on retries so blacklist / empty pools can escape.
            let forceDiversify = diversify || attempt > 0
            let session = try await APIClient.getSessionFeed(diversify: forceDiversify)
            lastSeedLabel = session.statusLabel
            seedStrategyLabel = session.statusLabel

            let nodes = rankedNodes(from: session.posts)

            guard let proposedCurrent = nodes.first else {
                continue
            }

            guard !proposedCurrent.isBlacklistedTopic else {
                continue
            }

            let remainingNodes = nodes
                .dropFirst()
                .filter { !$0.isBlacklistedTopic }

            return [proposedCurrent] + remainingNodes
        }

        seedStrategyLabel = lastSeedLabel
        throw GraphFeedError.noUsablePosts
    }

    private func loadChoices(for node: GraphFeedNode) async throws -> [GraphFeedNode] {
        let posts = try await APIClient.getNextGraphPosts(for: node.id)
        return rankedNodes(from: posts)
            .filter { choice in
                choice.id != node.id && !choice.isBlacklistedTopic
            }
    }

    private func nextAutomaticNode(excluding currentID: String) -> GraphFeedNode? {
        if let choice = nextChoices.first(where: { choice in
            choice.id != currentID && !choice.isBlacklistedTopic
        }) {
            nextChoices.removeAll { $0.id == choice.id || $0.isBlacklistedTopic }
            return choice
        }

        nextChoices.removeAll { $0.isBlacklistedTopic }

        while !sessionQueue.isEmpty {
            let nextNode = sessionQueue.removeFirst()
            if nextNode.id != currentID, !nextNode.isBlacklistedTopic {
                return nextNode
            }
        }

        return nil
    }

    private func rankedNodes(from posts: [Post]) -> [GraphFeedNode] {
        let uniquePosts = uniqued(posts)
        let nodes = uniquePosts.map(makeNode(from:))

        return nodes
            .enumerated()
            .sorted { lhs, rhs in
                if lhs.element.isPreferredTopic != rhs.element.isPreferredTopic {
                    return lhs.element.isPreferredTopic && !rhs.element.isPreferredTopic
                }

                return lhs.offset < rhs.offset
            }
            .map(\.element)
    }

    private func makeNode(from post: Post) -> GraphFeedNode {
        let normalizedTopic = normalizedTopicName(from: post.topic)

        return GraphFeedNode(
            post: post,
            isPreferredTopic: contains(normalizedTopic, in: preferences.preferredTopics),
            isBlacklistedTopic: contains(normalizedTopic, in: preferences.blacklistedTopics)
        )
    }

    private func refreshPreferenceFlags() {
        if let currentNode {
            self.currentNode = makeNode(from: currentNode.post)
        }

        nextChoices = nextChoices.map { makeNode(from: $0.post) }
        sessionQueue = sessionQueue.map { makeNode(from: $0.post) }
    }

    private func refreshLikedPostIDs() async {
        do {
            let interactions = try await APIClient.getMyInteractions()
            likedPostIDs = Set(
                interactions
                    .filter { $0.type == .like }
                    .map(\.postId)
            )
        } catch {
            // Keep the local set if trail history can't be loaded.
        }
    }

    private func recordInteraction(for node: GraphFeedNode, type: GraphInteractionType) async throws {
        let payload = GraphInteractionPayload(
            postId: node.id,
            type: type,
            viewTime: viewTime(at: Date())
        )
        try await APIClient.recordInteraction(payload)
    }

    private func viewTime(at date: Date) -> TimeInterval {
        guard let currentPostStartedAt else {
            return 0
        }

        return max(0, date.timeIntervalSince(currentPostStartedAt))
    }

    private func normalizedTopicName(from topic: String?) -> String? {
        guard let topic else {
            return nil
        }

        let normalized = TopicPreferenceList.normalizedTopic(from: topic)
        return normalized.isEmpty ? nil : normalized
    }

    private func contains(_ normalizedTopic: String?, in topics: [String]) -> Bool {
        guard let normalizedTopic else {
            return false
        }

        return topics.contains { topic in
            topic.caseInsensitiveCompare(normalizedTopic) == .orderedSame
        }
    }

    private func uniqued(_ posts: [Post]) -> [Post] {
        var seen = Set<String>()

        return posts.filter { post in
            seen.insert(post.id).inserted
        }
    }

    private func statusMessage(for node: GraphFeedNode, defaultMessage: String) -> String {
        var parts: [String] = []

        if node.isPreferredTopic, let topicName = node.topicName {
            parts.append("Preferred topic active: \(topicName)")
        } else if let topicName = node.topicName {
            parts.append("Current topic: \(topicName)")
        }

        if let seedStrategyLabel {
            parts.append(seedStrategyLabel)
        }

        if parts.isEmpty {
            return defaultMessage
        }

        return parts.joined(separator: " · ")
    }

    private func handle(_ error: Error, using authSession: AuthSession, fallbackMessage: String) {
        if case APIClientError.unauthorized = error {
            authSession.signOut()
        }

        let description = error.localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines)
        errorMessage = description.isEmpty ? fallbackMessage : description
    }
}

private enum GraphFeedError: LocalizedError {
    case noUsablePosts

    var errorDescription: String? {
        switch self {
        case .noUsablePosts:
            return "No session posts matched your current topic rules."
        }
    }
}

File: BubblerAndroid/app/src/main/java/com/bubbler/android/features/graph/GraphFeedViewModel.kt
package com.bubbler.android.features.graph

import androidx.lifecycle.ViewModel
import com.bubbler.android.core.auth.AuthSession
import com.bubbler.android.core.network.ApiException
import com.bubbler.android.data.model.GraphFeedNode
import com.bubbler.android.data.model.GraphInteractionPayload
import com.bubbler.android.data.model.GraphInteractionType
import com.bubbler.android.data.model.UserPreferences
import com.bubbler.android.data.repository.FeedRepository
import com.bubbler.android.data.repository.GraphRepository
import com.bubbler.android.data.repository.PostRepository
import com.bubbler.android.data.repository.PreferencesRepository
import com.bubbler.android.data.repository.UserRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.time.Duration
import java.time.Instant
import kotlin.math.max

/**
 * Graph walk session state — mirrors Swift `GraphFeedViewModel`.
 *
 * Behavior to preserve (see architecture.md):
 * - Up to three session retries; force diversify after the first failure
 * - Session queue fallback when a node has no usable neighbors
 * - View-time recorded on advance
 * - Client-side preferred/blacklist flag refresh on preference changes
 */
class GraphFeedViewModel(
    private val authSession: AuthSession,
    private val feedRepository: FeedRepository,
    private val graphRepository: GraphRepository,
    private val preferencesRepository: PreferencesRepository,
    private val userRepository: UserRepository,
    private val postRepository: PostRepository,
    private val clock: () -> Instant = { Instant.now() },
) : ViewModel() {
    private val _currentNode = MutableStateFlow<GraphFeedNode?>(null)
    val currentNode: StateFlow<GraphFeedNode?> = _currentNode.asStateFlow()

    private val _nextChoices = MutableStateFlow<List<GraphFeedNode>>(emptyList())
    val nextChoices: StateFlow<List<GraphFeedNode>> = _nextChoices.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isSubmitting = MutableStateFlow(false)
    val isSubmitting: StateFlow<Boolean> = _isSubmitting.asStateFlow()

    private val _seedStrategyLabel = MutableStateFlow<String?>(null)
    val seedStrategyLabel: StateFlow<String?> = _seedStrategyLabel.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _statusMessage = MutableStateFlow<String?>(null)
    val statusMessage: StateFlow<String?> = _statusMessage.asStateFlow()

    private var preferences: UserPreferences = UserPreferences.PLACEHOLDER
    private var sessionQueue: MutableList<GraphFeedNode> = mutableListOf()
    private var currentPostStartedAt: Instant? = null

    val hasCurrentPost: Boolean
        get() = _currentNode.value != null

    val currentTopicName: String
        get() = _currentNode.value?.topicName ?: "Topicless bubble"

    val hasCurrentTopic: Boolean
        get() = _currentNode.value?.topicName != null

    val isCurrentTopicPreferred: Boolean
        get() = _currentNode.value?.isPreferredTopic == true

    val isCurrentTopicBlacklisted: Boolean
        get() = _currentNode.value?.isBlacklistedTopic == true

    suspend fun load() {
        if (_isLoading.value) return
        loadSession(
            diversify = false,
            message = "Building your graph session.",
        )
    }

    suspend fun refreshSession() {
        loadSession(
            diversify = true,
            message = "Exploring a fresh path across topics.",
        )
    }

    suspend fun choose(node: GraphFeedNode) {
        advance(
            interactionType = GraphInteractionType.EXPLORE,
            explicitNextNode = node,
            fallbackMessage = "Following a connected post.",
        )
    }

    suspend fun skipCurrentPost() {
        advance(
            interactionType = GraphInteractionType.SKIP,
            explicitNextNode = null,
            fallbackMessage = "Skipping ahead to the next bubble.",
        )
    }

    fun updateCurrentPostContent(content: String) {
        val current = _currentNode.value ?: return
        val existing = current.post
        val updatedPost = existing.copy(content = content)
        _currentNode.value = GraphFeedNode(
            post = updatedPost,
            isPreferredTopic = current.isPreferredTopic,
            isBlacklistedTopic = current.isBlacklistedTopic,
        )
    }

    /** Reloads preferences after PostCard owns the prefs update. */
    suspend fun syncTopicPreferences() {
        try {
            preferences = preferencesRepository.getPreferences().sanitized()
            refreshPreferenceFlags()

            val current = _currentNode.value
            if (current != null && current.isBlacklistedTopic) {
                _nextChoices.value = emptyList()
                sessionQueue = mutableListOf()
                recordInteraction(current, GraphInteractionType.SKIP)
                loadSession(
                    diversify = true,
                    message = "Topic blacklisted. Exploring other bubbles.",
                )
            }
        } catch (error: Exception) {
            handle(error, fallbackMessage = "We couldn't refresh topic preferences.")
        }
    }

    suspend fun deleteCurrentPost() {
        val current = _currentNode.value ?: return

        _isSubmitting.value = true
        _errorMessage.value = null

        try {
            postRepository.deletePost(current.id)
            authSession.showSuccessMessage("Post deleted.")
            advanceAfterCurrentPostRemoved(successMessage = "Deleted your post and moved ahead.")
        } catch (error: Exception) {
            handle(error, fallbackMessage = "We couldn't delete that post.")
        }

        _isSubmitting.value = false
    }

    /** Called when PostCard already deleted the current post via the API. */
    suspend fun handleCurrentPostDeleted() {
        advanceAfterCurrentPostRemoved(successMessage = "Deleted your post and moved ahead.")
    }

    private suspend fun loadSession(
        diversify: Boolean = false,
        message: String,
    ) {
        _isLoading.value = true
        _errorMessage.value = null
        _statusMessage.value = message

        try {
            preferences = preferencesRepository.getPreferences().sanitized()

            val sessionNodes = fetchUsableSessionNodes(diversify = diversify)
            val firstNode = sessionNodes.firstOrNull()
                ?: throw GraphFeedError.NoUsablePosts

            sessionQueue = sessionNodes.drop(1).toMutableList()
            setCurrentNode(firstNode)

            if (_errorMessage.value == null) {
                val seedNote = _seedStrategyLabel.value?.let { " $it." }.orEmpty()
                _statusMessage.value = GraphFeedRanking.statusMessage(
                    node = firstNode,
                    seedStrategyLabel = _seedStrategyLabel.value,
                    defaultMessage = "Session ready.$seedNote",
                )
            }
        } catch (error: Exception) {
            _currentNode.value = null
            _nextChoices.value = emptyList()
            sessionQueue = mutableListOf()
            _seedStrategyLabel.value = null
            handle(error, fallbackMessage = "We couldn't build a graph session right now.")
        }

        _isLoading.value = false
    }

    private suspend fun advance(
        interactionType: GraphInteractionType,
        explicitNextNode: GraphFeedNode?,
        fallbackMessage: String,
    ) {
        val current = _currentNode.value
        if (current == null) {
            loadSession(
                diversify = false,
                message = "Building your graph session.",
            )
            return
        }

        _isSubmitting.value = true
        _errorMessage.value = null

        try {
            recordInteraction(current, interactionType)

            when {
                explicitNextNode != null -> {
                    setCurrentNode(explicitNextNode)
                    if (_errorMessage.value == null) {
                        _currentNode.value?.let { node ->
                            _statusMessage.value = GraphFeedRanking.statusMessage(
                                node = node,
                                seedStrategyLabel = _seedStrategyLabel.value,
                                defaultMessage = fallbackMessage,
                            )
                        }
                    }
                }
                else -> {
                    val nextNode = nextAutomaticNode(excludingCurrentId = current.id)
                    if (nextNode != null) {
                        setCurrentNode(nextNode)
                        if (_errorMessage.value == null) {
                            _currentNode.value?.let { node ->
                                _statusMessage.value = GraphFeedRanking.statusMessage(
                                    node = node,
                                    seedStrategyLabel = _seedStrategyLabel.value,
                                    defaultMessage = fallbackMessage,
                                )
                            }
                        }
                    } else {
                        loadSession(
                            diversify = true,
                            message = "Loading a fresh cross-topic path.",
                        )
                    }
                }
            }
        } catch (error: Exception) {
            handle(error, fallbackMessage = "We couldn't save that interaction.")
        }

        _isSubmitting.value = false
    }

    private suspend fun advanceAfterCurrentPostRemoved(successMessage: String) {
        val current = _currentNode.value ?: return

        _nextChoices.value = _nextChoices.value.filterNot { it.id == current.id }
        sessionQueue.removeAll { it.id == current.id }

        val nextNode = nextAutomaticNode(excludingCurrentId = current.id)
        if (nextNode != null) {
            setCurrentNode(nextNode)
            if (_errorMessage.value == null) {
                _statusMessage.value = successMessage
            }
        } else {
            loadSession(
                diversify = true,
                message = "Deleted your post. Exploring other bubbles.",
            )
        }
    }

    private suspend fun setCurrentNode(node: GraphFeedNode) {
        // Re-annotate from live preferences so a stale preview node can't carry old flags.
        val annotated = GraphFeedRanking.makeNode(node.post, preferences)
        if (annotated.isBlacklistedTopic) {
            loadSession(
                diversify = true,
                message = "That topic is blacklisted, so exploring other bubbles.",
            )
            return
        }

        _currentNode.value = annotated
        currentPostStartedAt = clock()

        try {
            _nextChoices.value = loadChoices(forNode = annotated)

            if (_nextChoices.value.isEmpty() && sessionQueue.isEmpty()) {
                _statusMessage.value =
                    "No connected posts were available, so the next action will pull a fresh session."
            }
        } catch (error: Exception) {
            _nextChoices.value = emptyList()
            handle(error, fallbackMessage = "We couldn't load the connected posts.")
        }
    }

    private suspend fun fetchUsableSessionNodes(
        diversify: Boolean,
        maxAttempts: Int = 3,
    ): List<GraphFeedNode> {
        var lastSeedLabel: String? = null

        repeat(maxAttempts) { attempt ->
            // Escalate to diversify on retries so blacklist / empty pools can escape.
            val forceDiversify = diversify || attempt > 0
            val session = feedRepository.getSessionFeed(diversify = forceDiversify)
            lastSeedLabel = session.statusLabel
            _seedStrategyLabel.value = session.statusLabel

            val nodes = GraphFeedRanking.rankedNodes(session.posts, preferences)
            val proposedCurrent = nodes.firstOrNull() ?: return@repeat
            if (proposedCurrent.isBlacklistedTopic) return@repeat

            val remainingNodes = nodes
                .drop(1)
                .filterNot { it.isBlacklistedTopic }

            return listOf(proposedCurrent) + remainingNodes
        }

        _seedStrategyLabel.value = lastSeedLabel
        throw GraphFeedError.NoUsablePosts
    }

    private suspend fun loadChoices(forNode: GraphFeedNode): List<GraphFeedNode> {
        val posts = graphRepository.getNextGraphPosts(forNode.id)
        return GraphFeedRanking.rankedNodes(posts, preferences)
            .filter { choice ->
                choice.id != forNode.id && !choice.isBlacklistedTopic
            }
    }

    private fun nextAutomaticNode(excludingCurrentId: String): GraphFeedNode? {
        val choice = _nextChoices.value.firstOrNull { candidate ->
            candidate.id != excludingCurrentId && !candidate.isBlacklistedTopic
        }
        if (choice != null) {
            _nextChoices.value = _nextChoices.value.filterNot {
                it.id == choice.id || it.isBlacklistedTopic
            }
            return choice
        }

        _nextChoices.value = _nextChoices.value.filterNot { it.isBlacklistedTopic }

        while (sessionQueue.isNotEmpty()) {
            val nextNode = sessionQueue.removeAt(0)
            if (nextNode.id != excludingCurrentId && !nextNode.isBlacklistedTopic) {
                return nextNode
            }
        }

        return null
    }

    private fun refreshPreferenceFlags() {
        _currentNode.value?.let { current ->
            _currentNode.value = GraphFeedRanking.makeNode(current.post, preferences)
        }

        _nextChoices.value = GraphFeedRanking.rankedNodes(
            posts = _nextChoices.value.map { it.post },
            preferences = preferences,
        )
        sessionQueue = sessionQueue
            .map { GraphFeedRanking.makeNode(it.post, preferences) }
            .toMutableList()
    }

    private suspend fun recordInteraction(node: GraphFeedNode, type: GraphInteractionType) {
        val payload = GraphInteractionPayload(
            postId = node.id,
            type = type,
            viewTime = viewTime(clock()),
        )
        userRepository.recordInteraction(payload)
    }

    private fun viewTime(at: Instant): Double {
        val started = currentPostStartedAt ?: return 0.0
        return max(0.0, Duration.between(started, at).toMillis() / 1_000.0)
    }

    private fun handle(error: Exception, fallbackMessage: String) {
        if (error is ApiException.Unauthorized) {
            authSession.signOut()
        }

        val description = error.message?.trim().orEmpty()
        _errorMessage.value = description.ifEmpty { fallbackMessage }
    }
}

private sealed class GraphFeedError(message: String) : Exception(message) {
    data object NoUsablePosts : GraphFeedError(
        "No session posts matched your current topic rules.",
    )
}
